### PR TITLE
supporting mounting proc with namespaces

### PR DESCRIPTION
**Pull request by motahan**

### CODE REVIEW

**Comment by ruoso** [comment]
Can you make this into a specific feature test, so we can disable it in older versions of Linux where this is not supported?

**Comment by motahan** [comment]
Do you mean an arg passed to the userchroot program (like --install-devices) or a build time -D flag?

**Comment by motahan** [comment]
Actually the current implementation using unshare, will require kernel version 3.8 or higher. Providing the CLONE_NEWPID flag to unshare is supported in 3.8+ per the documentation (http://man7.org/linux/man-pages/man2/unshare.2.html)

I'm going to switch to using clone instead of unshare + fork because the CLONE_NEWPID support in clone() requires 2.6.24+

**Comment by ruoso** [comment]
hmm... I was expecting us to mount /proc on the --install-devices command and unmount on the --uninstall-devices. But maybe if we mount it in the namespace it goes away as that process exits?

**Comment by ruoso** [comment]
After reading a bit more, I think you're right, the mount will just go away as the process leaves and the namespace disappears.

**Comment by ruoso** [comment]
One more thing, I think you should accept /proc to already exist, since more than one userchroot invocation could be made on the the same directory.

**Comment by motahan** [comment]
The code accepts having a /proc directory and it'll mount the new /proc over the old one. You don't mean that they can exist at the same time in the same chroot directory right?

**Comment by ruoso** [comment]
I meant the mkdir call that aborts everything if it fails...

**Comment by motahan** [comment]
it won't abort if the directory already exists. I check for EEXIST != errno

**Comment by ruoso** [comment]
ah, sorry, I missed that.

File: userchroot.c [hunk 1]
```diff
@@ -208,6 +226,98 @@ static void portable_clearenv() {
 #endif
 }
 
+struct epilogue_data {
+    uid_t target_user;
+    char** argv;
+    char** envp;
+};
+
+void epilogue(struct epilogue_data* d) {
```
Comment (ruoso) [inline userchroot.c:235]: I think I'd rather just pass the values individually instead of creating that struct...

Comment (motahan) [inline userchroot.c:235]: Me too but the clone child function only accepts one argument

Comment (ruoso) [inline userchroot.c:235]: oh, ok...

**Comment by motahan** [comment]
Any updates?

**Pull request merged**